server/src/main/java/org/elasticsearch/transport/ConnectionManager.java

```diff
@@ -122,40 +117,34 @@ public void connectToNode(DiscoveryNode node, ConnectionProfile connectionProfil
             return;
         }
 
-        try (Releasable lock = connectionLock.acquire(node.getId())) {
-            Transport.Connection connection = connectedNodes.get(node);
-            if (connection != null) {
-                assert connectingNodes.containsKey(node) == false;
-                lock.close();
-                connectingRefCounter.decRef();
-                listener.onResponse(null);
-                return;
-            }
+        if (pendingConnections.containsKey(node)) {
+            connectingRefCounter.decRef();
+            listener.onResponse(null);
+            return;
+        }
 
-            final List<ActionListener<Void>> connectionListeners = connectingNodes.computeIfAbsent(node, n -> new ArrayList<>());
-            connectionListeners.add(listener);
-            if (connectionListeners.size() > 1) {
-                // wait on previous entry to complete connection attempt
-                connectingRefCounter.decRef();
-                return;
-            }
+        final CompletableContext<Void> currentAttemptContext = new CompletableContext<>();
+        final CompletableContext<Void> existingContext = pendingConnections.putIfAbsent(node, currentAttemptContext);
+        if (existingContext != null) {
+            existingContext.addListener(ActionListener.toBiConsumer(listener));
+            // wait on previous entry to complete connection attempt
+            connectingRefCounter.decRef();
```

Comment:
let's put this in a `finally` to make sure this absolutely runs, even if `addListener` throws an exception